pkg/security/clientcert/cert_expiry_cache.go

```diff
@@ -0,0 +1,199 @@
+// time and the last time it was seen.
+type certInfo struct {
+	expiration int64
+	last_seen  time.Time
+}
+
+// CacheTTL is an overridable duration for when certificates should be evicted.
+var CacheTTL = 24 * time.Hour
+
+// PurgeLoopInterval is the interval at which the cache should be purged.
+var PurgeLoopInterval = time.Hour
+
+// The size of the cache and the certInfo struct. To be used for memory tracking.
+var GaugeSize = int64(unsafe.Sizeof(aggmetric.AggGauge{}))
+var CertInfoSize = int64(unsafe.Sizeof(certInfo{}))
+
+// Cache keeps track of when users certificates are expiring. It does this by
+// keeping track of every certificate which has been used within some recent
+// time window (24h) and groups the certificates by user. When metrics are
+// read, the cache reports for each user the earliest expiration for any of
+// their recently used certificates.
+type Cache struct {
+	mu      syncutil.Mutex
+	account MemAccount
+
+	cache             map[string]map[string]certInfo
+	expirationMetrics *aggmetric.AggGauge
+	stopper           *stop.Stopper
+	timesource        timeutil.TimeSource
+	reportTTL         bool
+}
+
+func NewCache(
+	account MemAccount,
+	stopper *stop.Stopper,
+	expirationMetrics *aggmetric.AggGauge,
+	timesource timeutil.TimeSource,
+	reportTTL bool,
+) *Cache {
+	cache := &Cache{
+		account: account,
+
+		cache:             make(map[string]map[string]certInfo),
+		expirationMetrics: expirationMetrics,
+		stopper:           stopper,
+		timesource:        timesource,
+		reportTTL:         reportTTL,
+	}
+	return cache
+}
+
+// Upsert updates the cache with the expiration time of a certificate.
+// If the user is not already in the cache, a gauge for its expiration is added.
+func (c *Cache) Upsert(ctx context.Context, user, serial string, expiration int64) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// if the user is not in the cache, add them, and add their expiration to a gauge.
+	if _, ok := c.cache[user]; !ok {
+		c.cache[user] = map[string]certInfo{}
+		c.expirationMetrics.AddFunctionalChild(func() int64 { return c.GetExpiration(user) }, user)
+
+		err := c.account.Grow(ctx, GaugeSize)
+		if err != nil {
+			return err
+		}
+	}
+
+	// insert / update the certificate expiration time.
+	certs := c.cache[user]
+	certs[serial] = certInfo{
+		expiration: expiration,
+		last_seen:  c.timesource.Now(),
+	}
+
+	return c.account.Grow(ctx, CertInfoSize)
+}
+
+// GetExpiration returns the earliest expiration time for any of the certificates
+// used by the given user. It returns 0 if the user is not in the cache. It also
+// returns the time-to-live of the earliest expiration in seconds if the cache
+// is configured to report on it via the `reportTTL` field.
+func (c *Cache) GetExpiration(user string) int64 {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if certs, ok := c.cache[user]; ok {
+		// compute the earliest expiration time.
+		var minExpiration int64
+		for _, cert := range certs {
```

Comment:
This does a scan through the certificates for a given user. It's assumed that a given user does not have a lot of certificates.